tests/filter/test_relative_dates.py

```diff
@@ -0,0 +1,171 @@
+# This file contains functional tests that would normally be written as
+# Cram-style tests. Without an easy way to mock the time in Cram tests (which is
+# necessary to test features that utilize the system time), an alternative is to
+# write equivalent tests using pytest+freezegun.
```

Comment:
> We can mock system time with linker preload libraries like [libfaketime](https://github.com/wolfcw/libfaketime).

I considered this but went against it since it requires extra setup for developers.

> Alternatively, we can [inject a tiny bit of code into Python's startup](https://gist.github.com/tsibley/ffb6469067b1cf8976be227d52548749)

Oh, this is a cool approach! Nice afternoon project 😄

I took it for a spin in 1aa4988f57785ddda358bb8e9b376f50eed68aaa. While it works ([with a slight caveat](https://github.com/nextstrain/augur/blob/1aa4988f57785ddda358bb8e9b376f50eed68aaa/tests/functional/filter/cram/filter-relative-dates.t#L5-L8)), I think the current pytest form is easier to author and maintain because of the parameterization feature. I only copied 4/12 tests from `test_filter_relative_dates` and the verbosity is noticeable in the Cram file.

Unless you know of a comparable way to parameterize both inputs and expected outputs in a Cram test, I think I'll update the comment at the top of this file to something along the lines of "pytest is good for parameterization + mocking system time" and leave other changes as-is.

Reply:
Sounds good. Totally fine to leave as-is.

> Unless you know of a comparable way to parameterize both inputs and expected outputs in a Cram test

Parameterizing inputs can be done in a loop or similar. For outputs, I think the best you can do is use the parameterized inputs to generate the expected outputs from a static template, and then let Cram compare the actual output to the templated output.

Or, stepping back, you could parameterize the whole Cram test file (e.g. the test file is the loop body, rather than containing the loop) and either run it with different env var values or statically generate the files based on externally-kept parameters.

But we don't have to use Cram for everything!

Reply:
Those are neat ideas! I'll save them for another day – I think we've made enough changes to Cram test authoring recently.

Reworded comment in [force-push](https://github.com/nextstrain/augur/compare/579c7e3e3b3a75fcdcc38765184a8706dedaa011..d7785a38cecf75068a5b48fda4958b7c7aeac749).